guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+can will receive. This would be useful if you had some service that had a hard
+upper limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of running requests
+        //    currently.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. So if the service is receiving a lot of requests we
+risk spending a lot of time waiting. It would be nice if we had a way to tell
+the caller that there was no more capacity left. Maybe they wanted to react to
+that by calling another service or somehow creating another service to handle
+part of the load.
+
+In a way it would be nice if `Service` had a method like this:
+
+```rust
+trait Service<R> {
+    async fn ready(&mut self);
+}
+```
+
+`ready` would be an async function that completes when the service has capacity
+enough to receive one new request. We would then require users to first call
+`service.ready().await` before doing `service.call(request).await`.
+
+Separating "calling the service" from "reserving capacity" also unlocks new use
+cases such as being able to maintain a set of "ready services" that we keep up
+to date in the background such that when a request arrives we already have a
+ready service to send it to and don't have to first wait for it to become ready.
+
+With this setup `ConcurrencyLimit` could track capacity inside `ready` and not
+allow users to call `call` until there is sufficient capacity.
+
+`Service`s that don't care about capacity can just return immediately from
+`ready` or delegate to some inner `Service` they're wrapping.
+
+However we still cannot define async functions inside traits. We could add
+another associated type to `Service` called `ReadyFuture` but having to return a
+`Future` will give us the same lifetime issues we've run into before. It would
+be nice if there was someway around that.
+
+Instead we can take some inspiration from the `Future` trait and define a method
```

Comment:
actually technically the inspiration came from `Sink` since it originally had a `start_send` but I think @carllerche wrote that too?

Reply:
Yeah totally fair, just thought I'd mention it xD